### PR TITLE
gcs handler implementation

### LINKED ISSUE #9856: Add GCP Object Storage integration
### Short description of current behavior

The goal of this issue is to build a new handler to integrate GCP Object Storage with MindsDB.

## Resources :bulb: 

:checkered_flag: To get started, please refer to the following resources:

* :books:  [Building Application Handler docs](https://docs.mindsdb.com/contribute/app-handlers): This documentation will guide you through the process of building a custom handler for MindsDB. It contains essential information on the handler structure, methods, and best practices.
* :book:  [GCP Docs](https://cloud.google.com/storage)
* :book:  [GCP Storage Python client](https://cloud.google.com/python/docs/reference/storage/latest) 
*  :point_right:  [Binance API Handler](https://github.com/mindsdb/mindsdb/tree/staging/mindsdb/integrations/handlers/binance_handler): Another example of a successful MindsDB integration to the Binance API to make trade data available to use for model training and predictions.

## Next Steps :male_detective: :female_detective: 
 
-   You can familiarize yourself with GCP Object Storage.
-   Create a new handler that connects MindsDB to GCP.
-   Implement the core features mentioned above.
-   Thoroughly test the integration to ensure proper functionality.
-   Submit a pull request with the new GCP Integration handler.

## :loudspeaker: Additional rewards :trophy:

Each App integration brings :eight:  :zero:  : points for SWAG and entry into the draw for a  :point_down: :
- :computer:  [Razer Blade 16 Laptop](https://www.razer.com/gaming-laptops/Razer-Blade-16/RZ09-0483SEJ3-R3U1)
- :shirt: :bear:  [MindsDB Swag](https://store.covver.io/mindsdb/collections/swag-collection)
:information_source: For more info check out https://mindsdb.com/hacktoberfest/ :point_left:

### PR DESCRIPTION
**Pull request by Better-Boy**
## Description

I've completed the new integration for Google Cloud Storage. Initial testing shows that it's functioning as expected. If you notice anything I've overlooked or any bugs, please let me know.

You can find the implementation details (using google-cloud-storage and duckdb) in the README.md file.

Fixes #9856 

## Type of change

(Please delete options that are not relevant)

- [ ] 🐛 Bug fix (non-breaking change which fixes an issue)
- [x] ⚡ New feature (non-breaking change which adds functionality)
- [ ] 📢 Breaking change (fix or feature that would cause existing functionality not to work as expected)
- [ ] 📄 This change requires a documentation update

## Verification Process

To ensure the changes are working as expected:

 - [ ]   Test Location: Specify the URL or path for testing.
 - [ ]   Verification Steps: Outline the steps or queries needed to validate the change. Include any data, configurations, or actions required to reproduce or see the new functionality.

## Additional Media:

- [x] I have attached a brief loom video or screenshots showcasing the new functionality or change.

1. Establishing connection
Screenshot 2024-10-16 at 1 26 30 AM

2. Querying TSV file from GCS
Screenshot 2024-10-16 at 1 29 16 AM

3. Querying CSV file from GCS
Screenshot 2024-10-16 at 1 29 44 AM

4. Querying Parquet file from GCS
Screenshot 2024-10-16 at 1 31 23 AM

5. Querying Json file from GCS
Screenshot 2024-10-16 at 1 33 37 AM

## Checklist:

- [x] My code follows the style guidelines(PEP 8) of MindsDB.
- [x] I have appropriately commented on my code, especially in complex areas.
- [x] Necessary documentation updates are either made or tracked in issues.
- [] Relevant unit and integration tests are updated or added.

### CODE REVIEW

File: mindsdb/integrations/handlers/gcs_handler/README.md [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+title: Google Cloud Storage
+sidebarTitle: Google Cloud Storage
+---
+
+This documentation describes the integration of MindsDB with [Google Cloud Storage](https://cloud.google.com/storage), an object storage service that offers industry-leading scalability, data availability, security, and performance.
+
+## Prerequisites
+
+1. Before proceeding, ensure that MindsDB is installed locally via [Docker](/setup/self-hosted/docker) or [Docker Desktop](/setup/self-hosted/docker-desktop).
+2. Install the dependencies from the `requirements.txt` file. Run the following command
```
Comment (MinuraPunchihewa) [inline mindsdb/integrations/handlers/gcs_handler/README.md:11]: Can you please replace this second step with what is used throughout our documentation? You can refer [this](https://docs.mindsdb.com/integrations/data-integrations/google-bigquery).

**Review by MinuraPunchihewa — CHANGES_REQUESTED** [review]
Hi @Better-Boy,
Overall, this looks pretty good, but I have a few comments:
1. I think we should make it possible to pass the content of the service account keys directly when creating the connection. Otherwise, we would not be able to use this integration in a cloud environment. You can refer the approach taken in the [Google BigQuery](https://docs.mindsdb.com/integrations/data-integrations/google-bigquery) integration. Please use the same parameter names as well.
2. Is there a need for the `project_id` parameter? I see that you are extracting it from the service account keys and I don't see the parameter used in the code anywhere. What do you think?
3. You may have noticed that we recently made some changes to the S3 integration by adding in a 'virtual' table called `Files`. Can you please add this in? This will be for one of our components required moving forward.

**Comment by MinuraPunchihewa** [comment]
@Better-Boy Please address the flake issues that are logged here as well: https://docs.mindsdb.com/integrations/data-integrations/google-bigquery

File: mindsdb/integrations/handlers/gcs_handler/README.md (hunk 1 above)
Comment (Better-Boy) [reply mindsdb/integrations/handlers/gcs_handler/README.md]: Done

**Comment by Better-Boy** [comment]
> Hi @Better-Boy, Overall, this looks pretty good, but I have a few comments:
> 
> 1. I think we should make it possible to pass the content of the service account keys directly when creating the connection. Otherwise, we would not be able to use this integration in a cloud environment. You can refer the approach taken in the [Google BigQuery](https://docs.mindsdb.com/integrations/data-integrations/google-bigquery) integration. Please use the same parameter names as well.
> 2. Is there a need for the `project_id` parameter? I see that you are extracting it from the service account keys and I don't see the parameter used in the code anywhere. What do you think?
> 3. You may have noticed that we recently made some changes to the S3 integration by adding in a 'virtual' table called `Files`. Can you please add this in? This will be for one of our components required moving forward.

1. Done
2. Removed project_id from parameters
3. Done

**Comment by Better-Boy** [comment]
> @Better-Boy Please address the flake issues that are logged here as well: https://docs.mindsdb.com/integrations/data-integrations/google-bigquery

Done

File: mindsdb/integrations/handlers/gcs_handler/requirements.txt [hunk 2]
```diff
@@ -0,0 +1,5 @@
+duckdb
+fsspec
+gcsfs
```
Comment (MinuraPunchihewa) [inline mindsdb/integrations/handlers/gcs_handler/requirements.txt:3]: I don't see this package utilized anywhere in the code. Is it required?

File: mindsdb/integrations/handlers/gcs_handler/gcs_handler.py [hunk 3]
```diff
@@ -0,0 +1,360 @@
...
+    RESPONSE_TYPE
+)
+
+from mindsdb.integrations.libs.api_handler import APIResource, APIHandler
+from mindsdb.integrations.utilities.sql_utils import FilterCondition
+
+logger = log.getLogger(__name__)
+
+
+class ListFilesTable(APIResource):
```
Comment (MinuraPunchihewa) [inline mindsdb/integrations/handlers/gcs_handler/gcs_handler.py:28]: Can we move the table classes to a different module called `gcs_tables.py`? That would make the code a little more modular.

File: mindsdb/integrations/handlers/gcs_handler/requirements.txt [hunk 4]
```diff
@@ -0,0 +1,5 @@
...
+google-auth
+google-cloud-storage
```
Comment (MinuraPunchihewa) [inline mindsdb/integrations/handlers/gcs_handler/requirements.txt:5]: In order for the use of these packages to be accepted by our CI checks, you will need to add them to this [dict](https://github.com/mindsdb/mindsdb/blob/da0e83985974d7796897c25b3c40ecf57c17e485/tests/scripts/check_requirements.py#L71). I think looking at the examples should allow you to understand how these are handled, but if you have any questions, let me know.

File: mindsdb/integrations/handlers/gcs_handler/gcs_handler.py [hunk 5]
```diff
@@ -0,0 +1,360 @@
...
+        Args:
+            name (Text): The name of the handler instance.
+            connection_data (Dict): The connection data required to connect to the GCS account.
+            kwargs: Arbitrary keyword arguments.
+        """
+        super().__init__(name)
+        self.connection_data = connection_data
+        self.kwargs = kwargs
+        self.is_select_query = False
+        self.table_name = None
```
Comment (MinuraPunchihewa) [inline mindsdb/integrations/handlers/gcs_handler/gcs_handler.py:85]: These instance do not seem to be in use anymore - can we please remove them?

File: mindsdb/integrations/handlers/gcs_handler/connection_args.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    },
+    service_account_keys={
+        'type': ARG_TYPE.PATH,
+        'description': 'Path to the service account JSON file',
+        'secret': True
+    },
+    service_account_json={
+        'type': ARG_TYPE.DICT,
+        'description': 'Content of service account JSON file',
+        'secret': True
```
Comment (MinuraPunchihewa) [inline mindsdb/integrations/handlers/gcs_handler/connection_args.py:20]: Can we add `label` for these arguments as well?

**Review by MinuraPunchihewa — CHANGES_REQUESTED** [review]

File: mindsdb/integrations/handlers/gcs_handler/requirements.txt (hunk 2 above)
Comment (Better-Boy) [reply mindsdb/integrations/handlers/gcs_handler/requirements.txt:3]: Yes, it will be used by fsspec.

Comment (MinuraPunchihewa) [reply mindsdb/integrations/handlers/gcs_handler/requirements.txt:3]: In that case, you will need to create another variable here similar to [this](https://github.com/mindsdb/mindsdb/blob/da0e83985974d7796897c25b3c40ecf57c17e485/tests/scripts/check_requirements.py#L64) and store the name of the package in there. After that, you will need to add it to this [list](https://github.com/mindsdb/mindsdb/blob/da0e83985974d7796897c25b3c40ecf57c17e485/tests/scripts/check_requirements.py#L67).

File: mindsdb/integrations/handlers/gcs_handler/gcs_handler.py (hunk 3 above)
Comment (Better-Boy) [reply mindsdb/integrations/handlers/gcs_handler/gcs_handler.py]: Done

File: mindsdb/integrations/handlers/gcs_handler/requirements.txt (hunk 4 above)
Comment (Better-Boy) [reply mindsdb/integrations/handlers/gcs_handler/requirements.txt:5]: Done

File: mindsdb/integrations/handlers/gcs_handler/gcs_handler.py (hunk 5 above)
Comment (Better-Boy) [reply mindsdb/integrations/handlers/gcs_handler/gcs_handler.py]: Removed it.

File: mindsdb/integrations/handlers/gcs_handler/connection_args.py (hunk 6 above)
Comment (Better-Boy) [reply mindsdb/integrations/handlers/gcs_handler/connection_args.py]: Done

File: mindsdb/integrations/handlers/gcs_handler/requirements.txt [hunk 7]
```diff
@@ -0,0 +1,5 @@
+duckdb
```
Comment (MinuraPunchihewa) [inline mindsdb/integrations/handlers/gcs_handler/requirements.txt:1]: Let's remove this from the list here as it is already part of the main dependencies for MindsDB. It is causing the CI to fail.

**Review by MinuraPunchihewa — CHANGES_REQUESTED** [review]

File: mindsdb/integrations/handlers/gcs_handler/requirements.txt (hunk 7 above)
Comment (Better-Boy) [reply mindsdb/integrations/handlers/gcs_handler/requirements.txt]: Done

**Review by MinuraPunchihewa — APPROVED** [review]

**Pull request merged**